AbstractSQLCompiler.ts

```diff
@@ -45,37 +45,48 @@ export interface Relationship {
 }
 export interface AbstractSqlQuery extends Array<AbstractSqlQuery | string> {}
 export interface AbstractSqlModel {
-	synonyms: {
-		[ synonym: string ]: string
-	}
-	relationships: {
-		[ resourceName: string ]: Relationship
-	}
-	tables: {
-		[ resourceName: string ]: AbstractSqlTable
-	}
+	synonyms: ResourceMap<string>
+	relationships: ResourceMap<Relationship>
+	tables: ResourceMap<AbstractSqlTable>
 	rules: AbstractSqlQuery[]
 }
 export interface SqlModel {
-	synonyms: {
-		[ synonym: string ]: string
-	}
-	relationships: {
-		[ resourceName: string ]: Relationship
-	}
-	tables: {
-		[ resourceName: string ]: AbstractSqlTable
-	}
+	synonyms: ResourceMap<string>
+	relationships: ResourceMap<Relationship>
+	tables: ResourceMap<AbstractSqlTable>
 	rules: SqlRule[]
 	createSchema: string[]
 	dropSchema: string[]
 }
 
+export interface hasDependants {
+	[dependant: string]: true
+}
+
+export interface schemaDependencyMap {
+	[tableName: string]: {
+		resourceName: string
+		primitive: AbstractSqlTable['primitive']
+		createSQL: string
+		dropSQL: string
+		depends: string[]
+	}
+}
+
+export interface ResourceMap<T> {
+	[ resourceName: string ]: T
+}
+
 export interface ModifiedFields {
 	table: string
 	fields?: {}[]
 }
 
+type ReduceFn<T> = (accumulator: any, element: T) => any
```

Comment:
I'd prefer this to be better typed, my guess is you can do that for your purposes as
```typescript
type ReduceFn<T> = (accumulator: T[], element: T) => T[]
```